### ISSUE TITLE
What exactly is staged prediction?

### ISSUE THREAD

**Issue by astrogilda** [stanfordmlgroup/ngboost#139]
Can you please explain what it is?

**Comment by astrogilda** [comment]
NVM figured it out--catboost introduced it, helps in visualization.

**Comment by hiyamgh** [comment]
Hello,

I hope you can share with us your thoughts. I came up with the following:

- one of ngboost's parameters is called **n_estimators**. This denotes the number of boosting iterations to fit. Lets say this number is 2000 (the default is 500 in ngboost, but in the experimentation sections, they used 2000)
- in **staged prediction**, we will have 2000 point predictions. Usually if we are doing point predictions alone, we will have a vector of predicted points with a length equivalent to the number of points in test/validation, but only one vector. So here this is still the case but not one but rather 2000 of these vectors. 
- in **staged probabilistic predictions**, same as above but rather than point predictions they are probabilistic predictions
- from those 2000 point predictions and probability distributions, I can get the parameter M as being the iteration (one of those 2000) that led to the minimal RMSE between the actual and predicted (point) predictions

@tonyduan  @alejandroschuler  Could you please confirm what I said above ?

**Comment by tonyduan** [comment]
@hiyamgh Yes this understanding of `staged_prediction` is correct.

I would emphasize that if you're going to tune the hyper-parameter _M_ by using staged prediction in this manner, then you should calculate performance on a **held-out validation set** as opposed to re-using training set.
